### PR TITLE
Add emails to API v2 [PLAT-960]

### PR DESCRIPTION
**Pull request by erinspace**
## Purpose

Add a user emails list and detail for user settings. These are only accessible to the currently logged in user for themselves.

## Changes
- add `/emails/` relationship to the user detail endpoint for currently logged in user
- `/users/<user_id>/settings/emails/` list endpoint
- `/users/<user_id>/settings/emails/<email_id>` detail endpoint

For confirmed emails, `<email_id>` is a hashid of the `Email` object's pk.
For unconfirmed emails, `<email_id>` is the token for the unconfirmed email. 

## QA Notes

This is all in API v2, no frontend changes. For testing via the API, here's what to note on the different endpoints that were changed:

### User Detail Page
- The `/emails/` relationship and detail URL should appear in the user detail page only if you're viewing your own user detail page

### Emails List
- The `/emails/` list includes all confirmed and unconfirmed emails
- When an unconfirmed email gets a new confirmation token, it should appear in the list with that token as the email ID. The old token URL will no longer be present and will no longer resolve.
- `POST`-ing to the endpoint can create a new unconfirmed email
    - This should send a confirmation email to the email address added

### Email Detail
- For confirmed emails, `<email_id>` is a hashid of the `Email` object's pk.
- For unconfirmed emails, `<email_id>` is the token for the unconfirmed email. 
- `PUT` to a detail endpoint you can update to your primary email iff it's a confirmed email
- You can delete both confirmed and non-confirmed emails with a `DELETE` request iff it's not your primary email

## Documentation

On the way!

## Side Effects

* @sloria will need to change `api/base/settings/local.py` to include `HASHIDS_SALT` with a random string

## Ticket

https://openscience.atlassian.net/browse/PLAT-960

### CODE REVIEW

File: api/users/serializers.py [hunk 1]
```diff
@@ -298,3 +304,62 @@ class UserAccountDeactivateSerializer(BaseAPISerializer):
...
+        try:
+            token = user.add_unconfirmed_email(address)
+            user.save()
+        except ValidationError as e:
+            raise rest_framework_exceptions.ValidationError(e.message)
+
+        return UserEmail(email_id=token, address=address, confirmed=False, primary=False)
+
+    def update(self, instance, validated_data):
+        save = False
```
Comment (caseyrollins) [inline api/users/serializers.py]: Drive-by: is this save variable is necessary? Couldn't this be: 
```
def update(self, instance, validated_data):
	user = self.context['request'].user
	primary = validated_data.get('primary', None)
	if primary and instance.confirmed:
		user.username = instance.address
		user.save()
	elif primary and not instance.confirmed:
		raise rest_framework_exceptions.ValidationError('You cannot set an unconfirmed email address as your primary email address.')
	return instance
```

File: api/base/utils.py [hunk 2]
```diff
@@ -29,6 +30,8 @@
 
 UPDATE_METHODS = ['PUT', 'PATCH']
 
+hashids = Hashids(alphabet='abcdefghijklmnopqrstuvwxyz')
```
Comment (sloria) [inline api/base/utils.py]: We should pass a salt so that the hashids aren't decodable by anyone except the app: https://github.com/davidaurelio/hashids-python#using-a-custom-salt .

The salt should be configurable in `api/base/settings/local.py`, and it should be made required here: https://github.com/CenterForOpenScience/osf.io/blob/826c1cfe3825af8e2f156d235084758c3785306c/api/base/settings/__init__.py#L27-L28

File: api/users/views.py [hunk 3]
```diff
@@ -655,3 +658,96 @@ def post(self, request, *args, **kwargs):
...
+
+    # Overrides RetrieveUpdateDestroyAPIView
+    def get_object(self):
+        email_id = self.kwargs['email_id']
+        user = self.get_user()
+
+        # check to see if it's a confirmed email with hashed id
+        decoded_id = hashids.decode(email_id)
+        if decoded_id:
+            try:
```
Comment (sloria) [inline api/users/views.py]: Nit: Ideally, `try` blocks should only contain the code that is expected to raise an error.

```python
try:
    email = user.emails.get(id=decoded_id[0])
except Email.DoesNotExist:
    email = None
else:
    primary = email.address == user.username
    address = email.address
    confirmed = True

```

File: api_tests/users/views/test_user_settings.py [hunk 4]
```diff
@@ -137,3 +139,253 @@ def test_exceed_throttle(self, mock_mail, app, user_one, url, payload):
...
+
+        # primary email detail
+        primary_email = Email.objects.get(address=user_one.username)
+        primary_hash = self.get_hashid(primary_email.id)
+        url = '/{}users/{}/settings/emails/{}/'.format(API_BASE, user_one._id, primary_hash)
+        res = app.get(url, auth=user_one.auth)
+        assert res.status_code == 200
+        assert res.json['data']['attributes']['primary'] is True
+
+    def test_addding_new_token_for_unconfirmed_email(self, app, user_one):
```
Comment (sloria) [inline api_tests/users/views/test_user_settings.py]: Nit: typo here

File: api_tests/users/views/test_user_settings.py [hunk 5]
```diff
@@ -137,3 +139,253 @@ def test_exceed_throttle(self, mock_mail, app, user_one, url, payload):
...
+        res = app.patch_json_api(url, payload, auth=user_one.auth)
+        assert res.status_code == 200
+        user_one.reload()
+        assert user_one.username == new_email.address
+
+        # test set primary not current user
+        res = app.patch_json_api(url, payload, auth=user_two.auth, expect_errors=True)
+        assert res.status_code == 403
+
+        # test set primary not confimred fails
```
Comment (sloria) [inline api_tests/users/views/test_user_settings.py]: Nit: typo here

File: api/users/views.py [hunk 6]
```diff
@@ -655,3 +658,96 @@ def post(self, request, *args, **kwargs):
...
+
+    # Overrides RetrieveUpdateDestroyAPIView
+    def perform_destroy(self, instance):
+        user = self.get_user()
+        email = instance.address
+        if instance.confirmed:
+            try:
+                user.remove_email(email)
+            except PermissionsError as e:
+                raise ValidationError(e.message)
```
Comment (sloria) [inline api/users/views.py]: Minor: To ease the transition to Python 3, you can change `e.message` to `e.args[0]`. Built-in `Exception` no longer has the `message` attribute.

File: api/users/serializers.py [hunk 7]
```diff
@@ -297,3 +305,61 @@ class UserAccountDeactivateSerializer(BaseAPISerializer):
...
+            raise Conflict('This user already has registered with the email address {}'.format(address))
+        try:
+            token = user.add_unconfirmed_email(address)
+            user.save()
+            if settings.CONFIRM_REGISTRATIONS_BY_EMAIL:
+                send_confirm_email(user, email=address)
+        except ValidationError as e:
+            raise rest_framework_exceptions.ValidationError(e.message)
+
+        return UserEmail(email_id=token, address=address, confirmed=False, primary=False)
```
[the remark this answers is not in the document]
Comment (erinspace) [reply api/users/serializers.py]: Ah so we had a discussion about this somewhere (probably on flowdock, that'll teach me to not have moved it to the ticket...) that sometimes there won't be an email ID available yet as unconfirmed emails don't yet exist as an object and so don't have an email id that a hashid can be created from. I think this is the case here as well, as this is creating an unconfirmed email.  This is [also using token in the list view](https://github.com/CenterForOpenScience/osf.io/pull/8620/files#diff-dd88c3f184d385cd8bef57c939c6d955R691) in the case of unconfirmed emails.

Tho maybe I am missing something, or we need another approach, let me know!

[the remark this answers is not in the document]
Comment (sloria) [reply api/users/serializers.py]: Ah, that's right. Didn't catch that this was dealing with unconfirmed emails. This is fine, then.
